pkg/server/server.go

```diff
@@ -274,9 +274,9 @@ func stageFiles(ctx context.Context, sc *Context, controlConfig *config.Control)
 	}
 	dataDir = filepath.Join(controlConfig.DataDir, "manifests")
 
-	dnsIPFamilyPolicy := "PreferDualStack"
-	if len(controlConfig.ClusterDNSs) == 1 {
-		dnsIPFamilyPolicy = "SingleStack"
+	dnsIPFamilyPolicy := "SingleStack"
+	if len(controlConfig.ClusterDNSs) > 1 {
+		dnsIPFamilyPolicy = "RequireDualStack"
```

Comment:
Thanks for openning the PR @aofei, I was wondering if we should add validation here, since the PR now will assume that if you assign another ip address to the dns it will be always considered as ipv6 and thats not necessarily the case always, its possible for the user to provide more than one ipv4 address for the clusterdns, so I think there some validation to see if one of the ips provided is ipv6  then in that case we should consider requireDualStack option for the dnsIPfamilyPolicy

Reply:
Hi @galal-hussein! I agree that k3s should do some validation to ensure that users are providing the correct IP families to kube-dns, but I don't think that should be done here. I believe such validation should be done [here](https://github.com/k3s-io/k3s/blob/a7fe1aaaa560234b03ae12756136bb2e77e40abc/pkg/cli/server/server.go#L339-L361) or [here](https://github.com/k3s-io/k3s/blob/a7fe1aaaa560234b03ae12756136bb2e77e40abc/pkg/cli/server/server.go#L551-L564).

Furthermore, considering that [`ClusterDNS`](https://github.com/k3s-io/k3s/blob/a7fe1aaaa560234b03ae12756136bb2e77e40abc/pkg/cli/server/server.go#L339-L361), [`ClusterIPRange`](https://github.com/k3s-io/k3s/blob/a7fe1aaaa560234b03ae12756136bb2e77e40abc/pkg/cli/server/server.go#L296-L310), and [`ServiceIPRanges`](https://github.com/k3s-io/k3s/blob/a7fe1aaaa560234b03ae12756136bb2e77e40abc/pkg/cli/server/server.go#L312-L325) face similar issues, it seems necessary to address these validation concerns in a separate PR.